### PR TITLE
Add "revision freeze" CI

### PR DESCRIPTION
**Pull request by sed-i**
## Issue
Charm metadata have major version "floating" resource tags such as [`ubuntu/prometheus:2-22.04`](https://github.com/canonical/prometheus-k8s-operator/blob/13a7202718f6441ff5c3101ccbba4cdbd633f00d/metadata.yaml#L83). This makes it difficult to track airgapped deployments.

## Solution
Add workflow for producing a manifest for airgapped deployments.

This CI could be called after successful matrix tests (related: #102).

In tandem with:
- https://github.com/canonical/charming-actions/pull/133

## Context
The output of this new CI (json manifest) could be then used for downloading charms and resources
- `juju info grafana-k8s --format=json | jq -r '.id'`
- `https://api.charmhub.io/api/v1/resources/download/charm_${charm_id}.${resource_name}_${resource_revision}` 

See also:
- https://discourse.charmhub.io/t/obtain-all-associated-revisions-via-stores-api/13527
- https://git.launchpad.net/~qch2012/+git/fce-lab-nebula/tree/scripts/offline/download-charm-resource.sh?h=3-nodes

### CODE REVIEW

**Comment by sed-i** [comment]
itest currently fails because of https://github.com/canonical/traefik-k8s-operator/pull/315.

**Comment by PietroPasotti** [comment]
would it make sense to make a release of the bundle and attach the frozen charm revisions to it as a resource? 
It could be a single zipfile, so that all that users have to do is:
- download the file
- unzip it
- run `setup.sh` (or whatever) --> bundle deployed.

**Comment by sed-i** [comment]
> would it make sense to make a release of the bundle and attach the frozen charm revisions to it as a resource? It could be a single zipfile, so that all that users have to do is:
> 
>     * download the file
> 
>     * unzip it
> 
>     * run `setup.sh` (or whatever) --> bundle deployed.

Deploying the bundle would theoretically be possible with
```bash
juju deploy <(wget -qO- https://raw.githubusercontent.com/.../frozen-bundle.json | frozen2bundle)
```
where `frozen2bundle` is a hypothetical script to be authored by the airgapped deployer.

We could attach it as a resource to the bundle itself when the need arises, but in the meanwhile I was thinking a PR in a similar idea to rocks.

**Comment by sed-i** [comment]
@lucabello could you please take a look at the newly created file `matrix-edge.yaml`?

The idea is to have independent CI jobs per channel:

```mermaid
graph TD

scheduled --> freeze-edge --> edge-quality-gate -->|on-success| publish-pinned/edge
scheduled --> freeze-beta --> beta-quality-gate -->|on-success| publish-pinned/beta
scheduled --> freeze-candidate --> candidate-quality-gate -->|on-success| publish-pinned/candidate
scheduled --> freeze-stable --> stable-quality-gate -->|on-success| publish-pinned/stable
```

Currently it the quality gate is the same matrix tests, but it doesn't have to be.
Wdyt?

File: .github/workflows/matrix-edge.yaml
Comment (sed-i) [inline .github/workflows/matrix-edge.yaml:1]: Seems like impossible to do nested matrix so probably would need to duplicate this yaml 3 times, one for each channel.

**Comment by lucabello** [comment]
I think to avoid duplication we should turn this into a reusable workflow (probably named something like `bundle-matrix.yaml` or something (to add in [canonical/observability](https://github.com/canonical/observability)).  
We could extract the risk track into a variable and use `{{ inputs.release-track }}` instead of `edge`., then have the the caller workflow (in this repo) probably be a matrix on those four inputs!

**Review by lucabello — APPROVED** [review]

**Review by Abuelodelanada — APPROVED** [review]

**Pull request merged**